src/folders/useDocumentWithLinks.ts

```diff
@@ -0,0 +1,75 @@
+import { AutomergeUrl } from "@automerge/automerge-repo";
+import { useEffect, useState } from "react";
+import { useDocuments } from "./useDocuments";
+
+/** This hook helps with loading a document which contains links to other documents.
+ *
+ *  Conceptually we want to follow links found in the document and materialize the contents
+ *  of the linked documents inside the given document. This process can be recursive too.
+ *
+ *  Notably, the type of the "raw" automerge document is not the same as the document with links
+ *  materialized, so we need to think in terms of those two types separately.
+ *
+ *  This hook is very flexible and applies to any document schema because the caller provides
+ *  arguments which govern its behavior:
+ *
+ *  @param rootUrl The URL of the root doc to load
+ *  @param findLinks A function which crawls the current doc (with links materialized) and returns
+ *    the list of all Automerge URLs which we want to load.
+ *  @param materializeLinks Given a raw document and a map of data we've loaded for linked docs,
+ *    construct a final document with links "materialized" into the data.
+ */
+export const useDocumentWithLinks = <TRawDoc, TDocWithLinksMaterialized>({
```

Comment:
What happens if one of the documents can't be loaded?

Reply:
good point, I suspect it will freeze -- will investigate